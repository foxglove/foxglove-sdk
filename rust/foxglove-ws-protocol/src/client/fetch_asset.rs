use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Fetch asset message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#fetch-asset>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "fetchAsset", rename_all = "camelCase")]
pub struct FetchAsset {
    /// Asset URI.
    pub uri: String,
    /// Request ID.
    pub request_id: u32,
}

impl JsonMessage for FetchAsset {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> FetchAsset {
        FetchAsset {
            uri: "package://foxglove/example.urdf".to_string(),
            request_id: 42,
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::FetchAsset(orig));
    }
}
