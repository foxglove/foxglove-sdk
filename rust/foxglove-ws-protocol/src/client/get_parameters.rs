use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Get parameters message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#get-parameters>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "getParameters", rename_all = "camelCase")]
pub struct GetParameters {
    /// Parameter names.
    pub parameter_names: Vec<String>,
    /// Request ID.
    pub id: Option<String>,
}

impl JsonMessage for GetParameters {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> GetParameters {
        GetParameters {
            parameter_names: vec!["param1".to_string(), "param2".to_string()],
            id: Some("request-123".to_string()),
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::GetParameters(orig));
    }
}
