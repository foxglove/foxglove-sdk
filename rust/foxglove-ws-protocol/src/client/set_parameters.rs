use serde::{Deserialize, Serialize};

use crate::parameter::Parameter;
use crate::JsonMessage;

/// Set parameters message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#set-parameters>
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "setParameters", rename_all = "camelCase")]
pub struct SetParameters {
    /// Parameters.
    pub parameters: Vec<Parameter>,
    /// Request ID.
    pub id: Option<String>,
}

impl JsonMessage for SetParameters {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;
    use crate::parameter::Parameter;

    use super::*;

    fn message() -> SetParameters {
        SetParameters {
            parameters: vec![
                Parameter::empty("empty"),
                Parameter::float64("f64", 1.23),
                Parameter::float64_array("f64[]", vec![1.23, 4.56]),
                Parameter::byte_array("byte[]", [0x10, 0x20, 0x30]),
                Parameter::bool("bool", true),
            ],
            id: Some("request-123".to_string()),
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::SetParameters(orig));
    }
}
