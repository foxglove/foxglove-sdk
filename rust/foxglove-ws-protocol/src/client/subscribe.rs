//! Subscribe message types.

use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Subscribe message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#subscribe>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "subscribe", rename_all = "camelCase")]
pub struct Subscribe {
    /// Subscriptions.
    pub subscriptions: Vec<Subscription>,
}

/// A subscription for a [`Subscribe`] message.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Subscription {
    /// Subscription ID.
    pub id: u32,
    /// Channel ID.
    pub channel_id: u64,
}

impl JsonMessage for Subscribe {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> Subscribe {
        Subscribe {
            subscriptions: vec![
                Subscription {
                    id: 1,
                    channel_id: 10,
                },
                Subscription {
                    id: 2,
                    channel_id: 20,
                },
            ],
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::Subscribe(orig));
    }
}
