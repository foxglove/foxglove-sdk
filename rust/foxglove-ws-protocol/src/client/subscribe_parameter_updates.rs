use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Subscribe parameter updates message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#subscribe-parameter-update>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(
    tag = "op",
    rename = "subscribeParameterUpdates",
    rename_all = "camelCase"
)]
pub struct SubscribeParameterUpdates {
    /// Parameter names.
    pub parameter_names: Vec<String>,
}

impl SubscribeParameterUpdates {
    /// Creates a new SubscribeParameterUpdates from an iterator of strings.
    pub fn new(names: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self {
            parameter_names: names.into_iter().map(Into::into).collect(),
        }
    }
}

impl JsonMessage for SubscribeParameterUpdates {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> SubscribeParameterUpdates {
        SubscribeParameterUpdates {
            parameter_names: vec!["param1".to_string(), "param2".to_string()],
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::SubscribeParameterUpdates(orig));
    }
}
