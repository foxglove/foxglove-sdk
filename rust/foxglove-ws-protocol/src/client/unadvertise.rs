use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Client unadvertise message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#client-unadvertise>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "unadvertise", rename_all = "camelCase")]
pub struct Unadvertise {
    /// Channel IDs.
    pub channel_ids: Vec<u32>,
}

impl JsonMessage for Unadvertise {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> Unadvertise {
        Unadvertise {
            channel_ids: vec![1, 2, 3],
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::Unadvertise(orig));
    }
}
