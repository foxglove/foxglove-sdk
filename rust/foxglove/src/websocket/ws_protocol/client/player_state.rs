use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Playing {
    Playing,
    Paused,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlayerTime {
    pub sec: u64,
    pub nsec: u64,
}

impl PlayerTime {
    pub fn new(sec: u64, nsec: u64) -> Self {
        Self { sec, nsec }
    }
}

/// Player state message.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "playerState", rename_all = "camelCase")]
pub struct PlayerState {
    /// Playing state
    pub playing: Playing,
    /// Playback speed
    pub playback_speed: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    /// Seek playback time
    pub seek_playback_time: Option<PlayerTime>,
    /// Previous playback time (will be defined if a seek has occurred)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub seek_previous_playback_time: Option<PlayerTime>,
}

impl JsonMessage for PlayerState {}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    #[test]
    fn test_encode_paused() {
        let message_paused = PlayerState {
            playing: Playing::Paused,
            playback_speed: 0.5,
            seek_previous_playback_time: None,
            seek_playback_time: Some(PlayerTime {
                sec: 123,
                nsec: 456,
            }),
        };
        insta::assert_json_snapshot!(message_paused);
    }

    #[test]
    fn test_encode_with_previous_time() {
        let message_with_previous_time = PlayerState {
            playing: Playing::Playing,
            playback_speed: 1.0,
            seek_previous_playback_time: Some(PlayerTime {
                sec: 1000000,
                nsec: 500000000,
            }),
            seek_playback_time: Some(PlayerTime {
                sec: 1000005,
                nsec: 0,
            }),
        };
        insta::assert_json_snapshot!(message_with_previous_time);
    }

    #[test]
    fn test_roundtrip() {
        let message = PlayerState {
            playing: Playing::Playing,
            playback_speed: 1.0,
            seek_previous_playback_time: None,
            seek_playback_time: Some(PlayerTime {
                sec: 1234567,
                nsec: 890123456,
            }),
        };

        let buf = message.to_string();
        let parsed = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(parsed, ClientMessage::PlayerState(message));
    }

    #[test]
    fn test_parse_json() {
        let json = r#"{"op":"playerState","playing":"playing","playbackSpeed":1.5,"seekPlaybackTime":{"sec":100,"nsec":500000000}}"#;
        let msg = ClientMessage::parse_json(json).unwrap();
        match msg {
            ClientMessage::PlayerState(state) => {
                assert_eq!(state.playing, Playing::Playing);
                assert_eq!(state.playback_speed, 1.5);
                assert_eq!(
                    state.seek_playback_time,
                    Some(PlayerTime {
                        sec: 100,
                        nsec: 500000000,
                    })
                );
                assert_eq!(state.seek_previous_playback_time, None);
            }
            _ => panic!("Expected PlayerState message"),
        }
    }

    #[test]
    fn test_parse_json_with_previous_time() {
        let json = r#"{"op":"playerState","playing":"paused","playbackSpeed":2.0,"seekPreviousPlaybackTime":{"sec":50,"nsec":250000000},"seekPlaybackTime":{"sec":100,"nsec":500000000}}"#;
        let msg = ClientMessage::parse_json(json).unwrap();
        match msg {
            ClientMessage::PlayerState(state) => {
                assert_eq!(state.playing, Playing::Paused);
                assert_eq!(state.playback_speed, 2.0);
                assert_eq!(
                    state.seek_playback_time,
                    Some(PlayerTime {
                        sec: 100,
                        nsec: 500000000,
                    })
                );
                assert_eq!(
                    state.seek_previous_playback_time,
                    Some(PlayerTime {
                        sec: 50,
                        nsec: 250000000,
                    })
                );
            }
            _ => panic!("Expected PlayerState message"),
        }
    }
}
