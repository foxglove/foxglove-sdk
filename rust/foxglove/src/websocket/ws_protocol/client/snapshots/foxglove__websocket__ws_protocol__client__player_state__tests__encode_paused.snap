---
source: rust/foxglove/src/websocket/ws_protocol/client/player_state.rs
assertion_line: 60
expression: message_paused
---
{
  "op": "playerState",
  "playing": "paused",
  "playbackSpeed": 0.5,
  "seekPlaybackTime": {
    "sec": 123,
    "nsec": 456
  }
}
